File: src/euler10.rs
mod numthry;
use numthry::Primes;

fn sum_primes_below(limit: u64) -> u64 {
	Primes::new()
		.take_while(|&x| x < limit)
		.fold(0, |acc, x| acc + x)
}

/// Find the sum of all the primes below two million.
fn main() {
	println!("{}", sum_primes_below(2_000_000));
}

/// The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
#[test]
fn test() {
	let expected = 17;
	let result = sum_primes_below(10);
	assert_eq!(result, expected);
}

File: c/src/euler10.c
//prepare: -lm

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>
#include "number_theory.c"

/*
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.
*/

size_t sum_sieve_of_erastosthenes(size_t limit){
	barray* primes = sieve_of_erastosthenes(limit);
	//2. Add up the primes we found.
	size_t sum = 0;
	for (size_t i = 2; i <= limit; i++){
		if (bget(primes, i)) {
			sum += i;
		}
	}
	bfree(primes);
	return sum;
}

int main(){
	assert(sum_sieve_of_erastosthenes(10) == 17);
	printf("%li\n", sum_sieve_of_erastosthenes(2000000));
	return 0;
}
